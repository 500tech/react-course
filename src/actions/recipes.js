import * as consts from '../consts/action-types';

export const addRecipe = (id, title, description) => ({
  type: consts.ADD_RECIPE,
  payload: {
    id,
    title: title.trim(),
    description: description.trim()
  }
});

export const toggleRecipe = (id) => ({
  type: consts.TOGGLE_RECIPE,
  payload: id
});

export const fetchRecipes = () => ({
  type: consts.API,
  payload: {
    url: 'recipes.json',
    success: consts.SET_RECIPES
  }
});