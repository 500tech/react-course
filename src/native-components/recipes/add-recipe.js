import React from 'react';
import { View, Text, StyleSheet } from 'react-native';

class AddRecipe extends React.Component {
  render() {
    return (
      <View>
        <Text>Add recipe</Text>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'space-between',
    alignItems: 'center',
    padding: 20
  },
  inputWrapper: {
    width: '100%'
  },
  inputContainer: {
    marginBottom: 30,
    width: '100%'
  },
  inputLabel: {
    color: '#0e749a',
    fontSize: 14,
    fontWeight: 'bold',
    marginBottom: 10
  },
  textInput: {
    borderColor: '#0e749a',
    borderWidth: 1,
    borderRadius: 4,
    paddingLeft: 10,
    paddingRight: 10,
    paddingTop: 5,
    paddingBottom: 5,
    width: '100%',
    fontSize: 14
  },
  button: {
    height: 50,
    width: 200,
    borderRadius: 5,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#0e749a'
  },
  buttonLabel: {
    fontSize: 14,
    color: 'white'
  }
});

export default AddRecipe;