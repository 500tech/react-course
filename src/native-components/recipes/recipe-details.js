import React from 'react';
import { TouchableOpacity, Text, StyleSheet } from 'react-native';
import PropTypes from 'prop-types';

const RecipeDetails = ({ recipe, unselectRecipe }) => (
  <TouchableOpacity style={ styles.container }
                    onPress={ unselectRecipe }>
    <Text style={ styles.title }>
      { recipe.title }
    </Text>

    <Text style={ styles.description }>
      { recipe.description }
    </Text>
  </TouchableOpacity>
);

RecipeDetails.propTypes = {
  recipe: PropTypes.object.isRequired,
  unselectRecipe: PropTypes.func.isRequired
};

const styles = StyleSheet.create({
  container: {
    padding: 20,
    alignItems: 'center'
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
    width: '100%',
    textAlign: 'center',
    marginBottom: 20
  },
  description: {
    fontSize: 14,
    marginBottom: 50
  }
});

export default RecipeDetails