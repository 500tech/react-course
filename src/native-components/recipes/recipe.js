import React from 'react';
import { TouchableOpacity, View, Text, StyleSheet, Image } from 'react-native';

const Recipe = ({ recipe }) => {
  return (
    <TouchableOpacity style={ styles.container }>
      <View style={ styles.recipeWrapper }>
        <Image source={require('../../assets/dish.png')}
               style={ styles.image }/>

        <Text style={ styles.title }>
          { recipe.title }
        </Text>
      </View>

      <View style={ styles.separator }/>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  container: {
    width: '100%',
    height: 50,
    alignItems: 'center',
    marginBottom: 20
  },
  recipeWrapper: {
    width: '100%',
    flexDirection: 'row'
  },
  image: {
    width: 30,
    height: 20,
    marginRight: 15
  },
  title: {
    fontWeight: '300',
    color: '#535353'
  },
  separator: {
    marginTop: 10,
    height: 1,
    backgroundColor: '#d2d2d2',
    width: '95%',
    alignSelf: 'flex-end'
  }
});

export default Recipe;