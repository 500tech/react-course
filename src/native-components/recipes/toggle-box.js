import React from 'react';
import { Animated, TouchableWithoutFeedback, StyleSheet } from 'react-native';

class ToggleBox extends React.Component {
  componentWillMount() {
    this.animation = new Animated.Value(this.props.active ? 100 : 0);

    this.borderRadius = this.animation.interpolate({
      inputRange: [0, 100],
      outputRange: [5, 120]
    });

    this.backgroundColor = this.animation.interpolate({
      inputRange: [0, 100],
      outputRange: ['transparent', '#d16d51']
    });

    this.buttonRotation = this.animation.interpolate({
      inputRange: [0, 50, 100],
      outputRange: ['0deg', '720deg', '0deg']
    });

    this.buttonHeight = this.animation.interpolate({
      inputRange: [0, 50, 100],
      outputRange: [50, 120, 50]
    });

    this.labelColor = this.animation.interpolate({
      inputRange: [0, 100],
      outputRange: ['#d16d51', 'white']
    });
  }

  toggle = () => {
    const { active } = this.props;

    Animated.timing(
      this.animation,
      { toValue: active ? 0 : 100, duration: 2000 }
    ).start(this.props.toggle);
  };

  render() {
    const { title } = this.props;

    const containerStyle = [
      styles.container,
      {
        height: this.buttonHeight,
        borderRadius: this.borderRadius,
        backgroundColor: this.backgroundColor,
        transform: [{ rotate: this.buttonRotation }]
      }
    ];

    const labelStyle = [
      styles.label,
      { color: this.labelColor }
    ];

    return (
      <TouchableWithoutFeedback onPress={ this.toggle }>
        <Animated.View style={ containerStyle }>
          <Animated.Text style={ labelStyle }>
            { title }
          </Animated.Text>
        </Animated.View>
      </TouchableWithoutFeedback>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    width: 120,
    borderColor: '#d16d51',
    borderWidth: 2,
    alignItems: 'center',
    justifyContent: 'center'
  },
  label: {
    fontSize: 14
  }
});

export default ToggleBox;