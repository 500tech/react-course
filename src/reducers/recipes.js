import * as consts from '../consts/action-types';

const initialState = [];

const reducer = (recipes = initialState, action) => {
  switch (action.type) {
    case consts.ADD_RECIPE:
      const newRecipe = Object.assign({}, action.payload, {
        favorite: false
      });

      return recipes.concat(newRecipe);

    case consts.TOGGLE_RECIPE:
      return recipes.map(recipe => recipe.id !== action.payload
        ? recipe
        : Object.assign({}, recipe, { favorite: !recipe.favorite }));

    case consts.SET_RECIPES:
      return action.payload;

    default:
      return recipes;
  }
};

export default reducer;