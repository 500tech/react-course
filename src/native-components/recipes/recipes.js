import React from 'react';
import { connect } from 'react-redux';
import { toggleRecipe, fetchRecipes } from '../../actions/recipes';
import { View, ScrollView, StyleSheet, RefreshControl } from 'react-native';

import Recipe from 'native/recipes/recipe';

class Recipes extends React.Component {
  state = {
    isLoading: false
  };

  componentDidMount() {
    this.props.fetchRecipes();
  }

  refreshList = () => {
    this.setState({ isLoading: true }, () => {
      this.props.fetchRecipes();

      this.setState({ isLoading: false });
    });
  };

  render() {
    const { recipes } = this.props;

    return (
      <View style={ styles.container }>
        <ScrollView style={{ width: '100%', flex: 1 }}
                    refreshControl={ <RefreshControl refreshing={ this.state.isLoading }
                                                     onRefresh={ this.refreshList }/> }>
          { recipes.map(recipe => <Recipe recipe={recipe} key={recipe.id}/>) }
        </ScrollView>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    paddingTop: 50,
    paddingBottom: 20,
    paddingLeft: 50,
    paddingRight: 50,
    backgroundColor: 'white',
    overflow: 'visible'
  },
  refreshButton: {
    marginTop: 10,
    height: 50,
    width: 200,
    borderRadius: 5,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#0e749a'
  },
  refreshButtonLabel: {
    fontSize: 14,
    color: 'white'
  }
});

const mapStateToProps = (state) => ({
  recipes: state.recipes
});

export default connect(mapStateToProps, {
  toggle: toggleRecipe,
  fetchRecipes
})(Recipes);