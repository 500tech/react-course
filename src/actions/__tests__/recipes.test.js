import * as actions from '../recipes';

describe('actions', () => {
  it('should create an action to add a recipe with snapshot', () => {
    expect(actions.addRecipe(1, 'title', 'desc')).toMatchSnapshot();
  });

  it('should create an action to add a recipe with trimmed title', () => {
    expect(actions.addRecipe(1, 'title', ' trimmed ')).toMatchSnapshot();
  });

  it('should create fetch action', () => {
    expect(actions.fetchRecipes()).toMatchSnapshot();
  });
});
