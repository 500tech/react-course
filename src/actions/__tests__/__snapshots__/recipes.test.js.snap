// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions should create an action to add a recipe with snapshot 1`] = `
Object {
  "payload": Object {
    "description": "desc",
    "id": 1,
    "title": "title",
  },
  "type": "ADD_RECIPE",
}
`;

exports[`actions should create an action to add a recipe with trimmed title 1`] = `
Object {
  "payload": Object {
    "description": "trimmed",
    "id": 1,
    "title": "title",
  },
  "type": "ADD_RECIPE",
}
`;

exports[`actions should create fetch action 1`] = `
Object {
  "payload": Object {
    "success": "SET_RECIPES",
    "url": "recipes.json",
  },
  "type": "API",
}
`;
